lib/crypto.js

```diff
@@ -661,6 +661,33 @@ function filterDuplicates(names) {
   }).sort();
 }
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+exports.timingSafeEqual = timingSafeEqual;
+function timingSafeEqual(a, b) {
+  if (!(a instanceof Buffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (!(b instanceof Buffer))
+    throw new TypeError('Second argument must be a Buffer');
+  var key = randomBytes(32);
+  var ah = new Hmac('sha256', key).update(a).digest();
+  var bh = new Hmac('sha256', key).update(b).digest();
+  // The second test, for a.equals(b), is just in case of the vanishingly small
+  // chance of a collision. It only fires if the digest comparison passes and so
+  // doesn't leak timing information.
+  return ah.equals(bh) && a.equals(b);
```

Comment:
> An attacker will have to constantly find hmac collisions (with a random key) to perform a timing attack, and even further the input that generates collision should be controllable every time. The point is - it is impossible to get past this hmac check enough times to gather any timing information [...]

Thanks Fedor, I don't think so. We can simply think of the hmac code as any other arbitrary program code which would be executed before the actual `a.equals(b)` comparison is done. That's how an attacker would see it.

The `hmac` check is no serious hurdle at all, because `hmac` can only be expected to provide the cryptographic guarantees for which it is designed, and `hmac` was never designed to be used as a *time-altering-function* before a non-timing-safe comparison. Otherwise it would be the same as saying that we can use a 256-bit encryption function before an unsafe comparison, hoping that someone would have to beat the 256-bit encryption before our timing would leak. Executing a timing attack is always non-trivial, but this actually makes it fairly easy. One almost couldn't ask for better.

Ben is definitely right.